sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -175,8 +179,24 @@ EventImplPtr Scheduler::addHostAccessor(Requirement *Req,
 }
 
 void Scheduler::releaseHostAccessor(Requirement *Req) {
-  Req->MBlockedCmd->MEnqueueStatus = EnqueueResultT::SyclEnqueueReady;
+  Command *const BlockedCmd = Req->MBlockedCmd;
+
   std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
+
+  assert(BlockedCmd && "Can't find appropriate command to unblock");
+
+  BlockedCmd->MEnqueueStatus = EnqueueResultT::SyclEnqueueReady;
+
+  enqueueLeavesOfReqUnlocked(Req);
+}
+
+void Scheduler::enqueueLeavesOfReq(const Requirement *const Req) {
+  std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
+  enqueueLeavesOfReqUnlocked(Req);
+}
+
+// static
+void Scheduler::enqueueLeavesOfReqUnlocked(const Requirement *const Req) {
   MemObjRecord *Record = Req->MSYCLMemObj->MRecord.get();
```

Comment:
An example I have, with a simplified excerpt below where q  and q2 are SYCL queues default constructed, segfaults on the line above. The reason is MRecord is 0x0.  See my comments on commands.cpp:237 for some thoughts about it

```cpp
  {
    buffer<float, 2> b_A{h_A.data(), range<2>{WIDTH, HEIGHT}};
    buffer<float, 2> b_B{h_B.data(), range<2>{WIDTH, HEIGHT}};
    buffer<float, 2> b_C{h_C.data(), range<2>{WIDTH, HEIGHT}};

    auto CG1 = [&](handler &h) {
      auto d_A = b_A.get_access<sycl::access::mode::read>(h);
      auto d_B = b_B.get_access<sycl::access::mode::read>(h);
      auto d_C = b_C.get_access<sycl::access::mode::read_write>(h);
      h.single_task<class foo>([=]() {
          d_C[0][0] = d_A[0][0] * d_B[0][0];
      });
    };

    auto CG2 = [&](handler &h) {
      auto d_A = b_A.get_access<sycl::access::mode::read>(h);
      auto d_B = b_B.get_access<sycl::access::mode::read>(h);
      auto d_C = b_C.get_access<sycl::access::mode::read_write>(h);

      h.codeplay_host_task([=] {
          std::cout << " Hello world from host task " << std::endl;
          });
    };

    q.submit(CG1);
    q2.submit(CG2);
  }
```

Reply:
This should be fixed now.

Reply:
Adding two more enqueues causes the test to deadlock:

```cpp
// previous code stays same as before
q.submit(CG1);    // A
q2.submit(CG2);  // B
q2.submit(CG1);  // C
q.submit(CG2);   // D
```
Deadlock occurs on  https://github.com/s-kanaev/llvm/blob/89bd48b4bbedcabe26f4726af701b16f1ed03579/sycl/source/detail/event_impl.cpp#L58 , while trying to submit line C above.

Reply:
This one is fixed now.
Added test.